### DISCUSSION TITLE
`condvar wait not supported` on wasm32-wasi target

### DISCUSSION THREAD

**Discussion by xtuc** [tokio-rs/tokio#5029] [General]
Hello,

I'm using the current thread runtime and targeting the `wasm32-wasi` target.
Since [tokio-1.21.0] I was expecting it to work, but I'm running into `condvar wait not supported`. Please see the minimal repro to demonstrate: [xtuc/tokio-wasm-repro]. 

Is that expected? Shouldn't Tokio take advantage of Wasi's [poll_oneoff]? Also please let me know if this should be an issue.

[tokio-1.21.0]: https://github.com/tokio-rs/tokio/releases/tag/tokio-1.21.0
[xtuc/tokio-wasm-repro]: https://github.com/xtuc/tokio-wasm-repro
[poll_oneoff]: https://github.com/WebAssembly/WASI/blob/main/phases/snapshot/docs.md#poll_oneoff

**Comment by Darksonn** [comment]
From the documentation:

> Note also that if the runtime becomes indefinitely idle, it will panic immediately instead of blocking forever. On platforms that don’t support time, this means that the runtime can never be idle in any way.

This is the situation that you've encountered. Though, the panic message could probably be improved.

**Reply by xtuc** [reply]
Wasi does support time, is that not implemented in Tokio's wasm support yet?

**Reply by Darksonn** [reply]
Time is supported on wasi, but you have no timers, so it is irrelevant. Let me quote the important part specifically:

> Note also that if the runtime becomes indefinitely idle, it will panic immediately instead of blocking forever.

**Reply by xtuc** [reply]
> Time is supported on wasi, but you have no timers

@Darksonn  please forgive my ignorance but isn't that what [poll_oneoff](https://github.com/WebAssembly/WASI/blob/main/phases/snapshot/docs.md#poll_oneoff) is enabling?

poll_oneoff subscriptions include:
> clock The time value of clock [subscription_clock::id](https://github.com/WebAssembly/WASI/blob/main/phases/snapshot/docs.md#subscription_clock.id) has reached timestamp [subscription_clock::timeout](https://github.com/WebAssembly/WASI/blob/main/phases/snapshot/docs.md#subscription_clock.timeout).

**Reply by Darksonn** [reply]
You misunderstand. I meant "there are no timers in your program", not "wasi does not support timers".
